shelley/chain-and-ledger/executable-spec/test/Generator/Utxo.hs

```diff
@@ -0,0 +1,143 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- The HasTrace instance relies on test generators and
+-- hence not situated with the LEDGER STS.
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Generator.Utxo
+  ( genTx
+  )
+  where
+
+import           Control.Monad (when)
+import qualified Data.Map.Strict as Map
+import           Data.Sequence (Seq)
+import           Data.Set (Set)
+import           Hedgehog (Gen)
+
+import qualified Hedgehog.Gen as Gen
+
+import           Cardano.Crypto.DSIGN.Mock (MockDSIGN)
+import           Cardano.Crypto.Hash (ShortHash)
+
+import           Coin (Coin (..))
+import           Generator.Core (findPayKeyPair, genInteger, genNatural, toAddr)
+import           Generator.Delegation (genDCerts)
+import           Ledger.Core (dom)
+import           LedgerState (pattern UTxOState)
+import           MockTypes (Addr, DCert, DPState, KeyPair, KeyPairs, Tx, TxBody, TxIn, TxOut, UTxO,
+                     UTxOState)
+import           Slot (Slot (..))
+import           STS.Ledger (LedgerEnv (..))
+import           Tx (pattern Tx, pattern TxBody, pattern TxOut)
+import           Updates (emptyUpdate)
+import           UTxO (pattern UTxO, balance, makeWitnessesVKey)
+
+-- | Generate a new transaction in the context of the LEDGER STS environment and state.
+--
+-- Selects unspent outputs and spends the funds on a some valid addresses.
+-- Also generates valid certificates.
+genTx :: LedgerEnv ShortHash MockDSIGN
```

Comment:
For generating transactions we should really really take a look at what @ruhatch did in `module Ledger.UTxO.Generators`. We need to think carefully in the shrink tree interleaves to make sure we get proper shrinking.

Reply:
I've deferred this bigger issue to #884